src/main/java/org/deepsymmetry/beatlink/VirtualRekordbox.java

```diff
@@ -368,13 +362,45 @@ private DeviceUpdate buildUpdate(DatagramPacket packet) {
                 byte[] data = packet.getData();
                 // PSSI Data
                 if (data[0x25] == 10) {
+
                     int rekordboxId = (int) Util.bytesToNumber(data, 0x28, 4);
-                    // If track is loaded
-                    if (rekordboxId != 0) {
+                    // If track is loaded and OpusProvider has attached media
+                    if (rekordboxId != 0 && OpusProvider.getInstance().hasAttachedArchive()) {
                         final byte[] pssiFromOpus = Arrays.copyOfRange(data, 0x35, data.length);
-                        final CdjStatus.TrackSourceSlot slot = (data[0x2d] == 3) ? SD_SLOT: USB_SLOT;
-                        final int player = data[0x21];
-                        OpusProvider.getInstance().handlePSSIMatching(rekordboxId, pssiFromOpus, player, slot);
+                        final int player = Util.translateOpusPlayerNumbers(data[0x21]);
+
+                        OpusProvider.getInstance().handlePSSIMatching(rekordboxId, pssiFromOpus, player);
+
+                        SlotReference slotRef = SlotReference.getSlotReference(player, USB_SLOT);
```

Comment:
So this is close but not quite there. The `MediaDetails` needs to report which player owns the slot the track lives in, _not_ which player is sending this update (i.e. playing the track). So we need to do the track matching _before_ we can figure out what slot would be used in building a `MediaDetails` object. But we should not be building one here in any case: the only time that one should be created and delivered as an update is when the user has attached a new metadata archive associated with a player number. Media updates are _rare_. This is sending them often. This does mean, though, that as things stand the constructor will need to be public despite my comment above because `OpusProvider` will be initiating the creation and delivery of those updates, and it is in a different package. So if I want to stop people calling it I’d need to modularize this project which I have avoided so far (modules are very anti-Clojure anyway). 😄

Reply:
We have an if statement a few lines down that will detect if MediaDetails already exists and skip sending if it does. Are you saying that MediaDetails just needs to all be coming from player 9 rather than one from each of the players?